### ISSUE TITLE
Example file does not compile with SILE v0.14 on macOS

### ISSUE THREAD

**Issue by bumbleblue** [sile-typesetter/sile#1837]
Using macOS Ventura 13.4.1 and installed the most recent SILE via homebrew.
I got the following error running `sile book.sil` in my local directory:

```  
SILE v0.14.10 (Lua 5.4)
<book.sil> as sil
</opt/homebrew/Cellar/sile/0.14.10/share/sile/core/sile.lua:413> as lua

! SILE.registerCommand() was deprecated in SILE v0.14.0. Please use class:registerCommand() instead.
Commands are being scoped to the document classes they are loaded into rather than being globals.

 at /opt/homebrew/Cellar/sile/0.14.10/share/sile/core/sile.lua:413: in <snippet>:
		[[␤␤SILE.registerCommand("book:chapterfont", function (options, content)␤  SILE.call("medskip")␤  SILE]] near book.sil:2:1: in \script

! SILE.registerCommand() was deprecated in SILE v0.14.0. Please use class:registerCommand() instead.
Commands are being scoped to the document classes they are loaded into rather than being globals.

 at /opt/homebrew/Cellar/sile/0.14.10/share/sile/core/sile.lua:413: in <snippet>:
		[[␤␤SILE.registerCommand("book:chapterfont", function (options, content)␤  SILE.call("medskip")␤  SILE]] near book.sil:2:1: in \script

! SILE.registerCommand() was deprecated in SILE v0.14.0. Please use class:registerCommand() instead.
Commands are being scoped to the document classes they are loaded into rather than being globals.

 at /opt/homebrew/Cellar/sile/0.14.10/share/sile/core/sile.lua:413: in <snippet>:
		[[␤␤SILE.registerCommand("book:chapterfont", function (options, content)␤  SILE.call("medskip")␤  SILE]] near book.sil:2:1: in \script

! Couldn't find face 'DejaVu Sans' at book.sil: in \font[family="DejaVu Sans",size="10%pw"] near 55:1: in \titlepage

error summary:
	Processing at: book.sil: in \font[family="DejaVu Sans",size="10%pw"] near 55:1: in \titlepage
	Using code at: ...brew/Cellar/sile/0.14.10/share/sile/shapers/harfbuzz.lua:82: Couldn't find face 'DejaVu Sans'

Run with --traceback for more detailed trace leading up to errors.
```

**Comment by alerque** [comment]
It looks like the error here is that you don't have one of the fonts expected and HarfBuzz is complaining. This might be a change on HarfBuzz's end in not falling back to some random system font, but I'm not sure.

Thanks for the details of the report. I'll do some experiments on the Linux side and see if I can come up with anything. I don't have access to a Mac to compare right now, but for this situation we should get similar results.

**Comment by Omikhleia** [comment]
Not sure which label to put on such issue (maybe it's macOS specific and has to do with the font manager there?)

**Comment by alerque** [comment]
Not sure about labeling either, either we need more labels or just don't label things that haven't / can't be properly triaged.

I just poked at this and I think you're right, the shaper is probably not the issue, the font loader is.

@bumbleblue can you report the output of these three commands or your system? I'm including mine below but I'm expecting yours to be different, especially the second one will probably just error but the output might still be useful.

```console
$ sile -e 'SU.dump(SILE.fontManager);os.exit(0)'
SILE v0.14.10 (LuaJIT 2.1.0-beta3)
{
  face = "function: 0x7f04b9a9f630",
  fontconfig = {
    _face = "function: 0x7f04b9ab6de0"
  }
}
$ sile -t -e 'SU.dump(SILE.fontManager:face({}));os.exit(0)'
SILE v0.14.10 (LuaJIT 2.1.0-beta3)
{
  family = "Iosevka Nerd Font",
  filename = "/usr/share/fonts/TTF/IosevkaNerdFont-Medium.ttf",
  fullname = "Iosevka NF Medium",
  index = 0,
  pointsize = 12
}
$ sile -e 'SU.dump(SILE.fontManager:face({family="Gentium Plus"}));os.exit(0)'
SILE v0.14.10 (LuaJIT 2.1.0-beta3)
{
  family = "Gentium Plus",
  filename = "/usr/share/fonts/gentium/GentiumPlus-Regular.ttf",
  fullname = "Gentium Plus",
  index = 0,
  pointsize = 12
}
```

**Comment by alerque** [comment]
Thanks for that output. It is not what I expected to see. First it suggests that you have both fontconfig AND macfonts at your disposal. I didn't realize that was a possible outcome of the build system but presumably everybody under Homebrew at least is ending up with that scenario. That might be something we look into actually leveraging, but it isn't super relevant to this issue.

Second I believe from skimming the code that if you have it, macfonts is  being used in preference to fontconfig.

Third it did not crash just *looking* for a font, presumably your crash comes when it tries to *use* the results of the font search that failed.

Fourth it says macfonts is doing something different than fontconfig on failing to find a font. Fontconfig aggressively starts looking for fallbacks and basically returns SOME font at least no matter what you search for. On a well configured system it might be a user's preferred replacement for a specific font or a sensible fallback of a similar family if one isn't found, or baring that any random font found on the system. But it always returns something. Macfonts appears to be returning an empty result possible with a specific signal saying it didn't find anything. And SILE does not appear to be handling that properly. That's a bug.

As a workaround you can probably tell it specifically to use fontconfig not macfonts at runtime: `sile --fontmanager=fontconfig ...`. Can you give that a shot and see if it works as expected?

**Comment by bumbleblue** [comment]
You are correct, changing the fontmanager did the trick! I still have a few warnings from compiling book.sil but the output is there and beautiful as always. Thank you so much!

**Comment by alerque** [comment]
I was reviewing this wondering whether there is really something I can do for v0.14.11, but I'm not sure there is. I don't have access to a Mac system right now to experiment, but the real "fix" here is to create our own fallback system of some kind, perhaps falling through to fontconfig if macfonts fails to return a usable font. The idea I guess would be to get *something* output even if we can't find fonts.

That being said I'm not so sure this is a bug. If anything perhaps our current Linux behavior is the bug: if a document calls out a specific font for something and our system doesn't have that font, why would it be considered a bug if we throw an error (as above) noting the font could not be found?

**Comment by bumbleblue** [comment]
Like you said, I think the expected behaviour here would be to compile the document will a fallback (system?) font, and then throw a warning that the font couldn't be found, with instructions to change the fontconfig. I am happy to continue testing pre-releases on my mac if that helps :)

**Comment by alerque** [comment]
Help me think through this here. What you are expected isn't the default behaviour of Macfonts. If what you're expecting SILE to do is monkey patch our use of Macfonts so that it behaves like Fontconfig, why are we even using Macfonts? What's the advantage? Allowing people to compile SILE without Fontconfig? If so why not at least for Homebrew where Fontconfig is easy to get ... why not just make it the default for Homebrew at least?

Is there anything I'm missing here? Is Macfonts going to find some Fonts on the system that Fontconfig can't see is a default setup? What are we gaining by having it as the default for macOS?

**Comment by davidchisnall** [comment]
I'm not sure if it's the same issue or not, but I'm having difficulty with non-default fonts on macOS:

```console
$ sile -e 'SU.dump(SILE.fontManager);os.exit(0)'
SILE v0.15.5 (LuaJIT 2.1.1731485912) [Rust]
{
  face = "function: 0x6000022d40c0",
  fontconfig = {
    _face = "function: 0x6000022c5800"
  },
  macfonts = {
    _face = "function: 0x6000022d4060"
  }
}
$ sile -t -e 'SU.dump(SILE.fontManager:face({}));os.exit(0)'
SILE v0.15.5 (LuaJIT 2.1.1731485912) [Rust]
{
  index = 0,
  pointsize = 12
}
$ sile -e 'SU.dump(SILE.fontManager:face({family="Source Code Pro"}));os.exit(0)'
SILE v0.15.5 (LuaJIT 2.1.1731485912) [Rust]
{
  family = "Source Code Pro",
  filename = "/Users/theraven/Library/Fonts/SourceCodePro-Black.ttf",
  index = 0,
  pointsize = 12
}
```

So it looks as if it's finding the Source Code Pro font, but then it errors like this:

```
! Couldn't find face 'Source Code Pro'
 at 896:93: in \font["family",family="Source Code Pro"]
Error: runtime error: ...rew/Cellar/sile/0.15.5_2/share/sile/shapers/harfbuzz.lua:95: 
stack traceback:
	[C]: in ?
	[C]: in function 'error'
	.../Cellar/sile/0.15.5_2/share/sile/core/utilities/init.lua:336: in function 'error'
	...rew/Cellar/sile/0.15.5_2/share/sile/shapers/harfbuzz.lua:95: in function 'callback'
	/opt/homebrew/Cellar/sile/0.15.5_2/share/sile/core/font.lua:162: in function 'cache'
	/opt/homebrew/Cellar/sile/0.15.5_2/share/sile/core/font.lua:70: in function </opt/homebrew/Cellar/sile/0.15.5_2/share/sile/core/font.lua:7>
	/opt/homebrew/Cellar/sile/0.15.5_2/share/sile/core/sile.lua:622: in function 'call'
	/opt/homebrew/Cellar/sile/0.15.5_2/share/sile/core/sile.lua:413: in function 'process'
	/opt/homebrew/Cellar/sile/0.15.5_2/share/sile/core/sile.lua:416: in function 'process'
	/opt/homebrew/Cellar/sile/0.15.5_2/share/sile/core/sile.lua:416: in function 'process'
	/opt/homebrew/Cellar/sile/0.15.5_2/share/sile/core/sile.lua:416: in function 'process'
	...
	...es/share/lua/5.1/sile/packages/resilient/styles/init.lua:364: in function 'styleForAlignment'
	...es/share/lua/5.1/sile/packages/resilient/styles/init.lua:526: in function <...es/share/lua/5.1/sile/packages/resilient/styles/init.lua:501>
	/opt/homebrew/Cellar/sile/0.15.5_2/share/sile/core/sile.lua:622: in function 'call'
	./lua_modules/share/lua/5.1/sile/classes/resilient/book.lua:494: in function 'process'
	...s/share/lua/5.1/sile/packages/resilient/headers/init.lua:72: in function 'func'
	...mebrew/Cellar/sile/0.15.5_2/share/sile/core/settings.lua:299: in function 'temporarily'
	/opt/homebrew/Cellar/sile/0.15.5_2/share/sile/core/sile.lua:556: in function 'typesetNaturally'
	...s/share/lua/5.1/sile/packages/resilient/headers/init.lua:57: in function 'outputHeader'
	./lua_modules/share/lua/5.1/sile/classes/resilient/book.lua:450: in function 'endPage'
	./lua_modules/share/lua/5.1/sile/silex/classes/base.lua:627: in function 'finish'
	/opt/homebrew/Cellar/sile/0.15.5_2/share/sile/core/sile.lua:698: in function </opt/homebrew/Cellar/sile/0.15.5_2/share/sile/core/sile.lua:697>
```

**Comment by Omikhleia** [comment]
> @davidchisnall your code snippet looks like it has a syntax error:

It's perhaps unrelated, @alerque, but for an XML input at least, `<font family="Yolo"/>` produces these "options" in the AST:

```
{
  "family",
  family = "Yolo"
}
```

... Which _would_ be serialized as above in an error log, `\font["family",family="Yolo"]`

**Comment by davidchisnall** [comment]
Yes, I'm using the XML input due to #2106.  SILE doesn't print the line that contained the error, it prints the SIL serialisation of the parsed AST and that frequently is not quite what the XML shows, as in this example.

**Comment by davidchisnall** [comment]
Using `Hack` instead of `Source Code Pro` does appear to work (it would be nice if SILE documented the default fonts - in my input I'd like to be able to support whatever the default fixed-width font is), so it isn't an XML parser problem.

**Comment by davidchisnall** [comment]
At least from my side, the XML I generate always has key-value pairs for attributes.

**Comment by Omikhleia** [comment]
BTW, as documented, LXP does it so one can track the attributes ordering.

**Comment by alerque** [comment]
Okay we don't care about attribute ordering (do we?), so I'm going to drop them from the AST if that looks okay to you. Even if it's never caused an actual problem because we only ever *use* named values out of the options table, it should be possible to directly compare the AST generated from XML vs. SIL inputs and get identical results. Also for debugging cases like this it just makes sense to only show the data we're actually using.

**Comment by Omikhleia** [comment]
> so I'm going to drop them from the AST if that looks okay to you.

We use an efficient SAX parser (LXP/Expat), and would try to hack over it for no real gain but decreased performances (likely)?
I'm not sure, sounds ugly to me. We could rather
 - fix our SIL-like dump /pretty-print to avoid showing these, for comparison purpose,
 - or switch to a different XML parser (C our Rust-based, ideally SAX-based too) or hack own own LXP-derived version

 Both would be cost-saving (performance wise and memory wise), rather than cost-inducing.

**Comment by alerque** [comment]
> it would be nice if SILE documented the default fonts - in my input I'd like to be able to support whatever the default fixed-width font is

I just added a note about this to the manual.

That's two side quests, but so far I still can't make heads or tails of your original issue.

Can you try running an MWE document specifically with the font debug flag on?

```Console
$ sile -d fonts -b debug - -o - <<< '<sile><font family="Source Code Pro">test</font></sile>'
```

**Comment by alerque** [comment]
Also @davidchisnall did you try running your actual document with `sile --fontmanager=fontconfig ...` as suggested for the original issue? That would be useful to know in confirming if you have the same bug or not.

**Comment by alerque** [comment]
Thanks for the effort narrowing this down. I still don't know where the issue is but that *should* be enough to scare it out of the weeds.

**Comment by alerque** [comment]
I figured this out. We expected macfonts to return an empty result when it didn't find anything, but it doesn't, it returns some tidbits about the query in a result table. Bah.

The solution is twofold:

1. Setup a fallback such that by default on macOS we query macfonts (if SILE is compiled with that option) but fallback to fontconfig if it doesn't find a match. This makes sense because the fontconfig loader will fallback to anything it can find and just warn if it isn't a match instead of leaving us hanging like the macfonts loader. If we want to promote missing font warnings to errors (a reasonable thing to do for some use cases) we should do it ourselves across the board instead of having some systems do it for us and others not.

1. Don't trust the results from macfonts blindly assuming a result is a font unless it actually has the font attributes we need.

Both as well as some other incidental stuff I bumped into while debugging this are coming soon to a release near you.
